### DISCUSSION TITLE
How do I map characters from the terminal?

### DISCUSSION THREAD

**Discussion by marklacas** [tabemann/zeptoforth#19] [Q&A] [answered]
Sadly my terminal program requires control-backspace to delete text in Zeptoforth.  I would like to map that to my Delete Key.  Also the arrow-up and arrow-down only work in the editor, not the command line.  Is there a way to remap these keys to match my serial terminal and computer?

**Comment by tabemann** [comment, marked as answer]
I would personally highly suggest using [zeptocom.js](https://tabemann.github.io/zeptocomjs/zeptocom.html), a Forth web serial terminal, with zeptoforth, if you have access to Chrome or Chromium (but note that web serial does not appear to work with Chromium on FreeBSD). (If you want to look at the source code for zeptocom.js, or you want to run it with a local web server, it is also [on GitHub](https://github.com/tabemann/zeptocomjs/).) zeptocom.js provides many features including history, history completion, and mass code upload support.

If you do not want to use zeptocom.js and rather want to use a terminal emulator like picocom or GNU screen, there is also the option of using zeptoforth's line editor, which is part of `full` builds of zeptoforth. For starters, you should be able to remap your backspace/delete key when using picocom or GNU screen. Also, the line editor provides the niceties one would expect with a line editor, including history. You can enable it by executing `enable-line` and disable it by executing `disable-line`. Note that instead of disabling it for a mass code upload you can put it into upload mode with F1 and exit out of upload mode with F2 at the start of a line. Note, however, that picocom and GNU screen are ill-suited to mass code uploads because they provide no baked-in support for line synchronization, unlike zeptocom.js, so you will need to use (rather tricky, error-prone, and slow) line delays when uploading code with them; rather, I would highly suggest one instead use `utils/codeload3.py`, which requires Python 3 and pySerial, to do mass code uploads.

Failing that, you might want to look at stty to change the settings for your terminal environment, if that is an option. You can remap your Delete key to $7F, which is expected by zeptoforth, with `stty erase ^\?`.

Last but not least, if you do not want to use picocom or GNU screen, and cannot use stty, but do not feel averse to recompiling and rebuilding zeptoforth, you can always modify the zeptoforth source, in which the use of $7F is hard-coded for delete. There are a few places to change `7F` to `08` and `delete` to `backspace`:

`cmp tos, #0x7F` on line 968 of `src/m0/outer.fs` in `_do_refill`
`cmp tos, #0x7F` on line 976 of `src/common/outer.fs` in `_do_refill`
`dup delete = if` on line 555 of `src/common/forth/line.fs` in `handle-upload`
`dup delete = if` on line 635 of `src/common/forth/line.fs` in `line-edit`
`dup delete = if` on line 762 of `src/common/forth/edit.fs` in `edit` (note here replace `delete` with `$08`)

Then follow the build instructions in `README.md` to get a new build containing these changes.

**Comment by tabemann** [comment]
Note that I have modified my post above, in case you already saw it in its partial form.

**Reply by marklacas** [reply]
Thanks so much!

**Reply by marklacas** [reply]
I sorted the problem out by using Serial. . . and app in the Apple App Store.  Works perfectly.

**Comment by marklacas** [comment]
I tried using zeptocom, but it threw errors in Chrome on my M1 Mac.  Three or four I believe.  I attempted to make some changes to get it working.  Upload works but not key entry.

**Comment by marklacas** [comment]
Finally figured out how to use ZeptoCom on my Mac M1.  Didn't realize that the line below the terminal is the entry.  All good now.

**Comment by Andy2No** [comment]
I've found that in TeraTerm (v4.106) I can use Keyboard, on the Setup menu.

The first two options are;

Transmit DEL by:
[  ] Backspace key
[  ] Delete key

It seems okay to just tick both of them.  This gets saved in the .INI file if you then use Save setup, on the Setup menu (which defaults to using the Program Files (x86) folder for TeraTerm).

That can then be loaded in again using Restore Setup, on the Setup menu, or by starting TeraTerm with a shortcut set to load that .INI file.  In Windows that means setting the target of the shortcut to be like this:

"C:\Program Files (x86)\teraterm\ttermpro.exe" /f=Zepto.INI
